libkbfs/semaphore_disk_limiter.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"sync"
+
+	"github.com/keybase/kbfs/kbfssync"
+	"golang.org/x/net/context"
+)
+
+// semaphoreDiskLimiter is an implementation of diskLimiter that uses
+// a semaphore.
+type semaphoreDiskLimiter struct {
+	s                    *kbfssync.Semaphore
+	availableByteDivisor int64
+	maxByteLimit         int64
+
+	lock sync.RWMutex
+	// byteLimit is min(maxByteLimit,
+	// (availableBytes + journalBytes) / availableByteDivisor).
+	byteLimit      int64
+	journalBytes   int64
+	availableBytes uint64
+}
+
+var _ diskLimiter = (*semaphoreDiskLimiter)(nil)
+
+func newSemaphoreDiskLimiter(
+	availableByteDivisor, maxByteLimit int64) *semaphoreDiskLimiter {
+	if availableByteDivisor < 1 {
+		panic("availableByteDivisor must be >= 1")
+	}
+	if maxByteLimit <= 0 {
+		panic("maxByteLimit must be > 0")
+	}
+
+	s := kbfssync.NewSemaphore()
+	s.Release(maxByteLimit)
+	return &semaphoreDiskLimiter{
+		s:                    s,
+		availableByteDivisor: availableByteDivisor,
+		maxByteLimit:         maxByteLimit,
+
+		byteLimit: maxByteLimit,
+	}
+}
+
+func (s *semaphoreDiskLimiter) getByteLimit() int64 {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.byteLimit
+}
+
+func (s *semaphoreDiskLimiter) recalculateLimitsLocked() {
+	// TODO: Check for overflow.
+	dividedAvailableBytes := (s.availableBytes + uint64(s.journalBytes)) /
+		uint64(s.availableByteDivisor)
+	var newByteLimit int64
+	if dividedAvailableBytes > uint64(s.maxByteLimit) {
+		newByteLimit = s.maxByteLimit
+	} else {
+		newByteLimit = int64(dividedAvailableBytes)
+	}
+
+	oldByteLimit := s.byteLimit
+	s.byteLimit = newByteLimit
+
+	if newByteLimit > oldByteLimit {
+		s.s.Release(newByteLimit - oldByteLimit)
+	} else if newByteLimit < oldByteLimit {
+		s.s.ForceAcquire(oldByteLimit - newByteLimit)
```

Comment:
I feel like it's bad to let this force the semaphore into negative territory, at least while `s.availableBytes > 0` (or some other small limit).  We don't want it to suddenly start failing writes just because some other application dumped a bunch of data onto disk, as long as there's some disk space left.

I think the best way to do this depends on whatever is coming in your backpressure PR.  What we want is a smoothish backpressure function, instead of a hard stop whenever we update `availableBytes`.  Maybe for now we can do something like this?

```go
n := s.s.ForceAcquire(oldByteLimit - newByteLimit)
if n < 0 {
    // If we had room before, make sure we still have some minimum room now.
    // But don't reclaim past what the old limit was.
    reclaim := 10*1024*1024 - n  // Try to keep at least 10MB of available space.
    if reclaim > (oldByteLimit - newByteLimit) {
        reclaim = oldByteLimit - newByteLimit
    }
    s.s.Release(reclaim)
    s.byteLimit += reclaim
}
```

(Maybe it's better to build this into `kbfssync.Semaphore` so it's not racy, since this way some writes could fail before the `Release()` call happens.  Or you could grab the lock for reading during `beforeBlockPut`.)